### ISSUE TITLE
Error with library index file after update.

### ISSUE THREAD

**Issue by lucas-mior** [OpenModelica/OpenModelica#13045]
### Description

When clicking in `File -> Manage Libraries -> Install Library`, it fails with the error:
```
[5] 16:20:41 Scripting Error
Library index file //.openmodelica/libraries/index.json doesn't exist.
```

### Steps to Reproduce

Click in `File -> Manage Libraries -> Install Library`

### Expected Behavior

Open the list of libraries available to install.

### Screenshots

![image](https://github.com/user-attachments/assets/b71c7cd9-256e-47cc-8322-936f04be0c8d)

### Version and OS

 - OpenModelica Version: `v1.24.0-cmake`
 - OS: Linux

### Additional Context

`MODELICAPATH` is set as `/home/user/.openmodelica/`

**Comment by casella** [comment]
> OpenModelica Version: `v1.24.0-cmake`

Do you mean you built this yourself? You should set MODELICAPATH to `~/.openmodelica/libraries/` for the package manager to work, it expects to see the top-level library directories right there.

**Comment by lucas-mior** [comment]
I changed MODELICAPATH as instructed, still the same error.

**Comment by adeas31** [comment]
@perost any hint?

**Comment by lucas-mior** [comment]
This bug does not happen when building as instructed in README instead of installing from the AUR. However, now another bug happens, will create an issue for it.

**Comment by lucas-mior** [comment]
What do you mean? The problem is that OMEdit is trying to open '/./whatever', which my user does not have the permissions to.

**Comment by casella** [comment]
@lucas-mior, the OMC `Package Manager` is supposed to store installed libraries for each user in the ~/.openmodelica/libraries/ directory, with MODELICAPATH pointing to it. Since it is below your home directory, you should have all permissions to read and write it. What I mean is that the library directories must actually be there, i.e., you should have installed them.

**Comment by lucas-mior** [comment]
I understand that. But as you can see from the error messages I included in the issues, the problem is that openmodelica is trying to load the library from another directory (which is not the one I have set in MODELICAPATH) . Furthermore, I can't even *open* the library manager to install any library. The error message in question:
![image](https://github.com/user-attachments/assets/349fdf74-1b2c-4e8d-ad71-9dfe4d9315ca)

Anyway, for some reason this problem only happens in the AUR build.

**Comment by perost** [comment]
Since I use Arch Linux myself I tried installing it from AUR, and I get the same issue. I'm also getting weird errors from OMEdit that I don't normally get:
```
[1] 14:17:58 Scripting Notification
Warning: Failed to set locale: 'en_US'

[2] 14:13:34 Scripting Notification
Detected locale "C" with character encoding "ANSI_X3.4-1968", which is not UTF-8.
Qt depends on a UTF-8 locale, and has switched to "C.UTF-8" instead.
If this causes problems, reconfigure your locale. See the locale(1) manual
for more information.
Detected locale "C" with character encoding "ANSI_X3.4-1968", which is not UTF-8.
Qt depends on a UTF-8 locale, and has switched to "C.UTF-8" instead.
If this causes problems, reconfigure your locale. See the locale(1) manual
for more information.

[3] 14:13:34 Scripting Notification
Automatically loaded package Complex 4.0.0 due to uses annotation from Modelica.

[4] 14:13:34 Scripting Notification
Automatically loaded package ModelicaServices 4.0.0 due to uses annotation from Modelica.

[5] 14:13:34 Scripting Error
Failed to parse model instance json for class OpenModelica with error unterminated object.

[6] 14:13:34 Scripting Error
Failed to parse model instance json for class ModelicaServices with error unterminated object.

[7] 14:13:34 Scripting Error
Failed to parse model instance json for class Complex with error unterminated object.

[8] 14:13:35 Scripting Error
Failed to parse model instance json for class Modelica with error unterminated object.
```
The first warning about `en_US` is normal since I don't have that locale but OMEdit (or Qt?) tries to it anyway. The second warning about `ANSI_X3.4-1968` is something I've never seen though, and not something I get with other Qt programs.

OMEdit does load the libraries I already have installed successfully but doesn't show the correct icons and stuff since it can't parse the JSON, possibly because the locale is messed up.

I don't know why installing it from AUR causes all these issues, the PKGBUILD doesn't seem to be doing anything weird. But I will investigate and see what I can find out.

**Comment by perost** [comment]
It seems the compiler works fine but OMEdit is completely messed up. OMEdit uses `getHomeDirectoryPath()` to get the user's home directory, but this returns `/` which is obviously wrong. But `getHomeDirectoryPath()` works just fine when called in a script or via the CLI inside OMEdit, and even when printing the return value in the compiler it's always right.

So as far as I can tell it's the communication between OMEdit and omc that's broken somehow. I'm guessing it's caused by the weird locale issues that Qt complains about, but I have no clue why it only happens with the AUR build.

**Comment by perost** [comment]
Actually, there's one major difference, the AUR build uses Qt 6 while I just build with the default which seems to still be Qt 5. @adeas31, is OMEdit supposed to work with Qt 6?

**Comment by adeas31** [comment]
> Actually, there's one major difference, the AUR build uses Qt 6 while I just build with the default which seems to still be Qt 5. @adeas31, is OMEdit supposed to work with Qt 6?

Yes. OMEdit should work fine with Qt6, I myself tested it on Windows.
AFAIK @jschueller added the Qt6 support and he is using Arch Linux. So it should be fine on Arch Linux as well.

**Comment by perost** [comment]
> > Actually, there's one major difference, the AUR build uses Qt 6 while I just build with the default which seems to still be Qt 5. @adeas31, is OMEdit supposed to work with Qt 6?
> 
> Yes. OMEdit should work fine with Qt6, I myself tested it on Windows. AFAIK @jschueller added the Qt6 support and he is using Arch Linux. So it should be fine on Arch Linux as well.

I tried it and it's definitely Qt6 that's the issue, so I guess there are still some issues to fix there.

**Comment by jschueller** [comment]
maybe I missed something, what does this line do ?
https://github.com/OpenModelica/OpenModelica/blob/master/OMEdit/OMEditLIB/Util/Utilities.cpp#L628

**Comment by adeas31** [comment]
> maybe I missed something in #12485, what does this line do ? https://github.com/OpenModelica/OpenModelica/blob/master/OMEdit/OMEditLIB/Util/Utilities.cpp#L628

This sets the codec for reading/writing the INI settings file. Since we used `UTF-8` so this line is not needed in Qt6 and setting files of older versions are compatible https://doc.qt.io/qt-6/qsettings.html#compatibility-with-older-qt-versions

**Comment by jschueller** [comment]
the locale issue should be independent though (probably LANG is incorreclty set, check your locale settings)

**Comment by perost** [comment]
Fixed with OpenModelica/OMBootstrapping#5.

The issue was that the Qt interface was using `QString::fromUtf8` to convert MetaModelica strings to Qt strings, which was missing a `char*` cast and ended up choosing the wrong overload in Qt6 such that only the first character of the string was returned. This was already fixed in the compiler, but not in the bootstrapping compiler that the CMake build uses to generate the Qt interface.

I still have the weird locale issue with `ANSI_X3.4-1968`, but I have `LANG=sv_SE.utf8` and everything looks correct. I don't have any locale issues with other applications either, only OMEdit. It seems to be related to WebEngine, but at least it doesn't seem to be causing any issues other than annoying messages.

**Comment by jschueller** [comment]
nice.
there are many other occurences of "QString::fromUtf8(MMC_STRINGDATA(...))", should they be changed too ?

**Comment by perost** [comment]
> nice. there are many other occurences of "QString::fromUtf8(MMC_STRINGDATA(...))", should they be changed too ?

Where?

**Comment by jschueller** [comment]
In OMCompiler, mostly in OMCompiler/Compiler/Script/OpenModelicaScriptingAPIQt.cpp:
```
OMCompiler/Compiler/Template/GenerateAPIFunctionsTpl.mo:1980:        a_postCall = Tpl.writeTok(a_postCall, Tpl.ST_STRING(" = QString::fromUtf8(MMC_STRINGDATA("));
OMCompiler/Compiler/Template/GenerateAPIFunctionsTpl.mo:2000:        a_postCall = Tpl.writeTok(a_postCall, Tpl.ST_STRING(" = QString::fromUtf8(MMC_STRINGDATA("));
OMCompiler/Compiler/boot/bootstrap-sources/build/GenerateAPIFunctionsTpl.c:406:#define _OMC_LIT162_data " = QString::fromUtf8(MMC_STRINGDATA("
OMCompiler/Compiler/boot/bomc/bootstrap-sources/build/GenerateAPIFunctionsTpl.c:406:#define _OMC_LIT162_data " = QString::fromUtf8(MMC_STRINGDATA("
OMCompiler/Compiler/Script/OpenModelicaScriptingAPIQt.cpp:23:    result = QString::fromUtf8(MMC_STRINGDATA(result_mm));
OMCompiler/Compiler/Script/OpenModelicaScriptingAPIQt.cpp:884:    result.newCref = QString::fromUtf8(MMC_STRINGDATA(out1_mm));
OMCompiler/Compiler/Script/OpenModelicaScriptingAPIQt.cpp:914:    result.contents = QString::fromUtf8(MMC_STRINGDATA(out1_mm));
OMCompiler/Compiler/Script/OpenModelicaScriptingAPIQt.cpp:944:    result.contents = QString::fromUtf8(MMC_STRINGDATA(out1_mm));
OMCompiler/Compiler/Script/OpenModelicaScriptingAPIQt.cpp:1061:    result.cref = QString::fromUtf8(MMC_STRINGDATA(out1_mm));
OMCompiler/Compiler/Script/OpenModelicaScriptingAPIQt.cpp:1175:    result.path = QString::fromUtf8(MMC_STRINGDATA(out1_mm));
OMCompiler/Compiler/Script/OpenModelicaScriptingAPIQt.cpp:1459:    result.contents = QString::fromUtf8(MMC_STRINGDATA(out1_mm));
OMCompiler/Compiler/Script/OpenModelicaScriptingAPIQt.cpp:2185:    result = QString::fromUtf8(MMC_STRINGDATA(result_mm));
OMCompiler/Compiler/Script/OpenModelicaScriptingAPIQt.cpp:2269:    result = QString::fromUtf8(MMC_STRINGDATA(result_mm));
OMCompiler/Compiler/Script/OpenModelicaScriptingAPIQt.cpp:2317:    result = QString::fromUtf8(MMC_STRINGDATA(result_mm));
OMCompiler/Compiler/Script/OpenModelicaScriptingAPIQt.cpp:2352:    result = QString::fromUtf8(MMC_STRINGDATA(result_mm));
...
```

**Comment by perost** [comment]
> In OMCompiler, mostly in OMCompiler/Compiler/Script/OpenModelicaScriptingAPIQt.cpp:

That file is generated by the code that I changed. If you're using the CMake build you may need to delete the `OMCompiler/Compiler/boot/bomc` folder for the change to take effect, otherwise you probably just need to check out the latest master with my fix.

**Comment by jschueller** [comment]
ah ok

**Comment by jschueller** [comment]
these is also one in omshell:
```
OMShell/OMShell/OMShellGUI/omcinteractiveenvironment.cpp:175:    error_ = 
QString::fromUtf8(MMC_STRINGDATA(reply_str));
```

**Comment by perost** [comment]
> these is also one in omshell:
> 
> ```
> OMShell/OMShell/OMShellGUI/omcinteractiveenvironment.cpp:175:    error_ = 
> QString::fromUtf8(MMC_STRINGDATA(reply_str));
> ```

That was fixed months ago in #12485, I'm not seeing any issues in the current master.

**Comment by jschueller** [comment]
yes, sorry my copy was outdated

**Comment by casella** [comment]
@lucas-mior please try again and close the ticket if it works for you.

**Comment by Hofsmo** [comment]
I had the same problem installing from AUR today.

**Comment by perost** [comment]
> Looking for openmodelica on [AUR](https://aur.archlinux.org/packages?O=0&SeB=nd&K=openmodelica&outdated=&SB=p&SO=d&PP=50&submit=Go) I see: ![immagine](https://private-user-images.githubusercontent.com/4233148/381614685-31a2461a-55a8-43cc-b118-19358e7f3dce.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzAzMDI2MTcsIm5iZiI6MTczMDMwMjMxNywicGF0aCI6Ii80MjMzMTQ4LzM4MTYxNDY4NS0zMWEyNDYxYS01NWE4LTQzY2MtYjExOC0xOTM1OGU3ZjNkY2UucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAzMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMzBUMTUzMTU3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OTZkOTA1ZDliNTQwNzE3YmQ4MDU4MmQ0ZDI4YmVjOGI0NWI3NjY5Yzg4YzhmMTNlODljZGM0NzYxNDc1ZDE0YyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.LoDcOVfk1knmyMMYqWm1u5FP8w-MdjZQXm1piA-jd0E)
> 
> Is there a repository for the development version? Otherwise, the fix will be available with 1.25.0. Or maybe we could get it fixed with 1.24.1.
> 
> @perost, do I miss something?

It should just work since our CMake build downloads OMBootstrapping master for the bootstrapping sources, so any fixes there will be retroactively applied to older versions when using CMake (whether we want it or not). I just tried building the AUR package and it works just fine for me.

> I had the same problem installing from AUR today.

Did you use a clean folder or did you already have a src directory for the package checked out? If you already have the sources checked out you'll need to delete them first (or at least the `OMCompiler/Compiler/boot/bomc` folder), since the CMake build doesn't download new bootstrapping sources if they already exist.

**Comment by casella** [comment]
Sorry, I missed that the bootstraping sources are out of the release game.

**Comment by Hofsmo** [comment]
> > I had the same problem installing from AUR today.
> 
> Did you use a clean folder or did you already have a src directory for the package checked out? If you already have the sources checked out you'll need to delete them first (or at least the `OMCompiler/Compiler/boot/bomc` folder), since the CMake build doesn't download new bootstrapping sources if they already exist.

No, I did `git clean -dfx`, which deleted everything except the src folder. I retried today after deleting the src folder, which didn't work. However, it seems that it may have been a misunderstanding from my side. I went to OMShell and used the package manager to install the Modelica library, and now everything works. I am used to the Modelica library being available when I install OpenModelica, but I may have been spoiled from before. Many thanks for the help :) and sorry for my misunderstanding.

**Comment by adeas31** [comment]
@Hofsmo the question is where are the libraries installed? Are they installed in `/home/user/.openmodelica/libraries` and loads fine in OMEdit? Because this is what @perost fixed. His fix makes sure that OMEdit get correct user home instead of just `/`.

**Comment by Hofsmo** [comment]
> @Hofsmo the question is where are the libraries installed? Are they installed in `/home/user/.openmodelica/libraries` and loads fine in OMEdit? Because this is what @perost fixed. His fix makes sure that OMEdit get correct user home instead of just `/`.

They are installed in `/home/user/.openmodelica/libraries` after I used OMShell to install them. To begin with I also had the problem that OMEdit looked in `/`, which was fixed after applying the fix of @perost . However, after getting his fix I also had to use OMShell to install the libraries, which was unexpected for me. Although, that was probably my mistake.

**Comment by lucas-mior** [comment]
@casella I completely forgot that issue, sorry. But I can't even test it because now the AUR build just segfaults before opening OMEdit. So I am building from github (cmake build), and it works fine.
